File: message-io/src/main/java/net/easyrpc/message/io/MessageNode.java
package net.easyrpc.message.io;

import java.io.Closeable;
import java.io.IOException;
import java.util.concurrent.ConcurrentSkipListSet;
import java.util.concurrent.TimeUnit;

/***
 * MessageNode(core)
 * <li>A message node can define its behaviours using register method</li>
 * <li>A message node can bind some ports of localhost using listen method</li>
 * <li>A message node can connect it to some nodes using connect method</li>
 * <li>A message node is a keep-alive transport container</li>
 * <li>Message nodes are closeable</li>
 */
public interface MessageNode extends Connectible, Closeable {

    long ACCEPT_POLLING = 10;
    TimeUnit ACCEPT_POLLING_UNIT = TimeUnit.MILLISECONDS;

    long MESSAGE_POLLING = 100;
    TimeUnit MESSAGE_POLLING_UNIT = TimeUnit.MICROSECONDS;

    /***
     * get all transports
     *
     * @return connected transport
     */
    ConcurrentSkipListSet<Transport> transports();

    /***
     * register a message model
     *
     * @param tag     model tag
     * @param handler message model logic
     * @return this instance
     */
    <T> MessageNode register(String tag, Class<T> type, MessageHandler<T> handler);

    /***
     * add event model when transport connect
     *
     * @param handler transport model
     * @return this instance
     */
    MessageNode setConnectHandler(TransportHandler handler);

    /***
     * add event model when transport disconnect
     *
     * @param handler transport model
     * @return this instance
     */
    MessageNode setDisconnectHandler(TransportHandler handler);

    /***
     * open a listen port at localhost
     *
     * @param port listen port
     * @return this instance
     */
    MessageNode listen(int port) throws IOException;

    /***
     * static: create a new message actor instance
     *
     * @return message actor instance
     */
    static MessageNode create() {
        return new MessageNodeImpl();
    }
}

File: message-io/src/main/kotlin/net/easyrpc/message/io/api/MessageNode.kt
package net.easyrpc.message.io.api

import net.easyrpc.message.io.core.Transport
import java.io.Closeable
import java.io.IOException
import java.util.concurrent.ConcurrentSkipListSet

interface MessageNode : Closeable {

    /***
     * 获得当前建立的所有连接
     */
    fun transports(): ConcurrentSkipListSet<Transport>

    /***
     * 定义消息行为
     */
    fun <T> register(tag: String, type: Class<T>, handler: TypedMessageHandler<T>): MessageNode

    /***
     * 连接到另外的MessageNode
     */
    fun connect(host: String, port: Int, success: (Transport) -> Unit = {},
                fail: (IOException) -> Unit = {}): MessageNode

    /***
     * 断开已经建立的连接
     */
    fun disconnect(transport: Transport)

    /***
     * 监听本地端口
     */
    fun listen(port: Int, accept: (Transport) -> Unit = {},
               fail: (IOException) -> Unit = {}): MessageNode

    /***
     * 关闭本地端口
     */
    fun shutdown(port: Int)

}